src/studio/src/designer/DataModeling/Converter/Xml/XmlSchemaToJsonSchemaConverter.cs

```diff
@@ -1100,71 +1100,75 @@ private static void HandleType(XmlQualifiedName typeName, decimal minOccurs, dec
+            }
 
-                if (type != null)
-                {
-                    if (nillable)
-                    {
-                        typeBuilder.Type(new SchemaValueType[] { type.Value, SchemaValueType.Null });
-                    }
-                    else
-                    {
-                        typeBuilder.Type(type.Value);
-                    }
-                }
-                else
-                {
-                    if (nillable)
-                    {
-                        var refSchemaBuilder = new JsonSchemaBuilder();
-                        refSchemaBuilder.Ref(GetReferenceFromTypename(typeName));
+            if (array)
+            {
+                // Sets new builder for items node in array.
+                typeBuilder = new JsonSchemaBuilder();
+            }
 
-                        var typeSchemaBuilder = new JsonSchemaBuilder();
-                        typeSchemaBuilder.Type(SchemaValueType.Null);
+            if (xsdType != null)
+            {
+                typeBuilder.XsdType(xsdType);
+            }
 
-                        typeBuilder.OneOf(refSchemaBuilder, typeSchemaBuilder);
-                    }
-                    else
-                    {
-                        typeBuilder.Ref(GetReferenceFromTypename(typeName));
-                    }
-                }
+            if (type != null)
+            {
+                typeBuilder.Type(type.Value, nillable);
+            }
+            else
+            {
+                BuildReference(typeBuilder, typeName, nillable);
+            }
 
-                if (format != null)
-                {
-                    typeBuilder.Format(format);
-                }
+            if (format != null)
+            {
+                typeBuilder.Format(format);
+            }
 
-                if (array)
-                {
-                    if (minOccurs > 0)
-                    {
-                        builder.MinItems((uint)minOccurs);
-                    }
+            if (!array)
+            {
+                return;
+            }
 
-                    if (maxOccurs > 1 && maxOccurs < decimal.MaxValue)
-                    {
-                        builder.MaxItems((uint)maxOccurs);
-                    }
+            if (minOccurs > 0)
+            {
+                builder.MinItems((uint)minOccurs);
+            }
 
-                    JsonSchema itemsSchema = typeBuilder;
-                    typeBuilder = builder;
-                    builder.Type(SchemaValueType.Array);
-                    builder.Items(itemsSchema);
-                }
+            if (maxOccurs is > 1 and < decimal.MaxValue)
+            {
+                builder.MaxItems((uint)maxOccurs);
+            }
 
-                if (xsdType != null)
-                {
-                    typeBuilder.XsdType(xsdType);
-                }
+            builder.Type(SchemaValueType.Array);
+            builder.Items(typeBuilder);
```

Comment:
And this is where I get lost if it's the same reference:)

Reply:
It's never the same reference. This code will hit only arrays
There is a check above
```c#
if (!array)
{
    return;
}
```
Just the nesting is removed.
For arrays we've built type and xsdType in new schema and added that as part of Items keword.
But when I see diff in github it's impossible to follow since it's huge refactor.